src/fifolock.jl

```diff
@@ -0,0 +1,143 @@
+# the wait queue, transfer ownership to it, schedule it, and
+# return. Thus when one or more tasks are waiting,`havelock`
+# is never reset.
+
+"""
+    trylock(l::FIFOLock)
+
+Try to acquire lock `l`. If successful, return `true`. If the lock is
+held by another task, do not wait and return `false`. 
+"""
+@inline function trylock(l::FIFOLock)
+    ct = current_task()
+    if l.locked_by === ct
+        l.reentrancy_cnt += 0x0000_0001
+        return true
+    end
+    return _trylock(l, ct)
+end
+@noinline function _trylock(l::FIFOLock, ct::Task)
+    GC.disable_finalizers()
+    if (@atomicreplace :acquire l.havelock 0x00 => LOCKED_BIT).success
+        l.reentrancy_cnt = 0x0000_0001
+        @atomic :release l.locked_by = ct
+        return true
+    end
+    GC.enable_finalizers()
+    return false
+end
+
+"""
+    lock(l::FIFOLock)
+
+Acquire lock `l`. If the calling task has already acquired the lock
+previously, increment an internal counter and return to support
+reentrancy. Each `lock` call must be matched with an `unlock` call.
+
+As with `Base.ReentrantLock`, acquiring a lock will inhibit running
+finalizers on that thread until the lock is released.
+
+FIFO behavior is handled in `unlock`.
+"""
+@inline function lock(l::FIFOLock)
+    trylock(l) || _lock(l)
+end
+@noinline function _lock(l::FIFOLock)
+    ct = current_task()
+    c = l.cond_wait
+    lock(c)
+    try
+        _trylock(l, ct) && return
+        wait(c)
+        l.reentrancy_cnt = 0x0000_0001
+        @atomic :release l.locked_by = ct
+    finally
+        unlock(c)
+    end
+    return
+end
+
+"""
+    unlock(l::FIFOLock)
+
+Release ownership of the lock `l`. If the lock was acquired recursively,
+the number of unlocks must match the number of locks before `l` is
+actually released.
+
+FIFO behavior is enforced here, in `unlock`: if one or more tasks are
+waiting on the lock, we do not actually unlock; just hand ownership to
+the first waiting task and schedule it.
+"""
+@inline function unlock(l::FIFOLock)
+    ct = current_task()
+    if l.locked_by !== ct
+        error("unlock from wrong thread")
+    end
+    if l.reentrancy_cnt == 0x0000_0000
+        error("unlock count must match lock count")
+    end
+    _unlock(l)
+end
+@noinline function _unlock(l::FIFOLock)
+    ct = current_task()
+    c = l.cond_wait
+    n = l.reentrancy_cnt - 0x0000_0001
+    if n == 0x0000_0000
+        lock(c)
+        try
+            if isempty(c.waitq)
+                l.reentrancy_cnt = n
+                @atomic :release l.locked_by = nothing
+                @atomic :release l.havelock = 0x00
+            else
+                t = popfirst!(c.waitq)
+                @atomic :release l.locked_by = t
+                schedule(t)
+            end
+            GC.enable_finalizers()
+        finally
+            unlock(c)
+        end
+    else
+        l.reentrancy_cnt = n
+    end
+    return
+end
+
+const Condition = Base.GenericCondition{FIFOLock}
```

Comment:
Reusing the name `Condition` doesn't seem right. Can you name this `FIFOCondition`?